File: VIF_git/GSDS-Qual/LeetCode/week1/Linked_List/Maximum_Twin_Sum_of_a_Linked_List/inseo.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* start;
    int ans = 0;
    
    void help(ListNode* node){
        
        if(node == NULL)
            return;
        help(node->next);
        ans = max(ans, start->val+node->val);
        start = start->next;
    }
    int pairSum(ListNode* head) {
        start = head;   
        help(head);
        return ans;
    }
};

File: VIF_git/GSDS-Qual/LeetCode/week1/Linked_List/Maximum_Twin_Sum_of_a_Linked_List/inseo.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def pairSum(self, head: Optional[ListNode]) -> int:
        cur = head

        n = 0
        val_list = []
        while cur != None:
            val_list.append(cur.val)
            cur = cur.next
            n += 1

        answer = 0
        for i in range(int(n/2)):
            answer = max(answer, val_list[i]+val_list[n-i-1])
        return answer